raft/log.go

```diff
@@ -1750,35 +1768,43 @@ func (l *Log) initWriter(w io.Writer, id, term, index uint64) (*logWriter, error
 
 	// Wrap writer and append to log to tail.
 	writer := &logWriter{
-		Writer:        w,
-		id:            id,
-		snapshotIndex: l.FSM.Index(),
-		done:          make(chan struct{}),
+		Writer: w,
+		id:     id,
+		done:   make(chan struct{}),
 	}
 	l.writers = append(l.writers, writer)
 
+	// If index is before first available entry, then require snapshot.
+	if len(l.entries) == 0 || index < l.entries[0].Index {
+		writer.snapshotIndex = l.FSM.Index()
```

Comment:
It's the index that the writer should start from once the snapshot is done and it begins catching up. It uses the `FSM.Index()` since the follower will be starting from that index once the FSM is copied (aka snapshotted).